applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -307,57 +306,39 @@ def InitializeSolutionStep(self):
         self.primal_analysis.RunSolutionLoop()
         Logger.PrintInfo("> Time needed for solving the primal analysis = ",round(timer.time() - startTime,2),"s")
 
-        # TODO the response value calculation for stresses currently only works on the adjoint modelpart
-        # this needs to be improved, also the response value should be calculated on the PRIMAL modelpart!!
-        self.adjoint_analysis.time = self.adjoint_analysis._GetSolver().AdvanceInTime(self.adjoint_analysis.time)
-
-        # synchronize the modelparts
-        self._SynchronizeAdjointFromPrimal()
-
-        self.adjoint_analysis.InitializeSolutionStep()
-
-
     def CalculateValue(self):
         startTime = timer.time()
         value = self._GetResponseFunctionUtility().CalculateValue(self.primal_model_part)
         Logger.PrintInfo("> Time needed for calculating the response value = ",round(timer.time() - startTime,2),"s")
 
         self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE] = value
 
-
     def CalculateGradient(self):
-        Logger.PrintInfo("\n> Starting adjoint analysis for response:", self.identifier)
+        # synchronize the modelparts
+        self._SynchronizeAdjointFromPrimal()
         startTime = timer.time()
-        self.adjoint_analysis._GetSolver().Predict()
-        self.adjoint_analysis._GetSolver().SolveSolutionStep()
+        Logger.PrintInfo("\n> Starting adjoint analysis for response:", self.identifier)
+        if not self.adjoint_analysis.time < self.adjoint_analysis.end_time:
+            self.adjoint_analysis.end_time += 1
```

Comment:
hm this looks like quite a specific hack
Why is it needed?

Reply:
To be honest I just copied it from the primal problem above. I can remove it. @armingeiser is there a reason why you do this time check for the primal analysis?

Reply:
It is, but as Martin said, it is already done like this for the primal analysis. It is not related to this change.